File: avx512-utf8-to-utf32/validate/avx512-validate-leading-bytes.py
# proof for AVX512 approach

def length(nibble):
    if nibble & 0x8 == 0x0:
        return 1
    if nibble & 0xc == 0x8:
        return 0
    if nibble & 0xe == 0xc:
        return 2
    if nibble & 0xf == 0xe:
        return 3
    if nibble & 0xf == 0xf:
        return 4

    assert False


def generate_words():
    continuation = b'\x80'
    for byte0 in range(0, 256):
        if length(byte0 >> 4) < 2:
            continue

        for byte1 in range(256):
            if byte1 & 0xc0 != 0x80:
                continue

            b0 = byte0.to_bytes(1, 'little')
            b1 = byte1.to_bytes(1, 'little')

            char2 = b0 + b1
            v, r = is_valid_utf8(b0 + b1)
            if v:
                yield (r, v, byte0, byte1)
                continue

            v, r = is_valid_utf8(b0 + b1 + continuation)
            if v:
                yield (r, v, byte0, byte1)
                continue

            v, r = is_valid_utf8(b0 + b1 + continuation + continuation)
            if v:
                yield (r, v, byte0, byte1)
                continue

            yield (r, False, byte0, byte1)


def is_valid_utf8(bytes):
    try:
        str(bytes, encoding='utf8')
        return True, ''
    except UnicodeError as e:
        return False, str(e)


def algorithm(leading_byte, continuation1):
    assert 0 <= leading_byte <= 255
    assert 0 <= continuation1 <= 255

    nibble1 = leading_byte >> 4
    nibble0 = leading_byte & 0xf

    # pshufb of (leading_byte >> 4)
    n = length(nibble1)

    if n == 0:
        return (False, "cont")
    
    if n == 1:
        return (True, "ASCII")

    if n == 2:
        greater = (leading_byte >= 0xc2)
        return (greater, "len2")

    if False:
        # naive version
        if n == 3:
            M = (continuation1 & 0x3f) > 0x1f
            if nibble0 == 0b0000:
                return (M, "len3: M")
            if nibble0 == 0b1101:
                return (not M, "len3: !M")

            return (True, "len3")
    else:
        if n == 3:
            tmp = continuation1 & 0x3f
            if nibble0 == 0b0000:
                tmp = tmp ^ 0b0000_0000
            elif nibble0 == 0b1101:
                tmp = tmp ^ 0b0010_0000
            else:
                tmp = tmp ^ 0b1100_0000

            return (tmp > 0x1f, "len3")

    if False:
        if n == 4:
            M = (continuation1 & 0x3f) > 0x0f
            if nibble0 == 0b0000:
                return (M, f"len4: M ({M}, {continuation1:08b})")
            if nibble0 in (0b0001, 0b0010, 0b0011):
                return True, "len4"
            if nibble0 == 0b0100:
                return (not M, "len4: !M")
            
            return (False, "len4*")
    else:
        if n == 4:
            tmp = continuation1 & 0x3f

            a = (tmp - 0x10) & 0xff

            tmp = a ^ 0x40

            if nibble0 == 0b0000:
                # tmp > 0x0f
                mask = 0x40
            elif nibble0 in (0b0001, 0b0010, 0b0011):
                # true
                mask = 0xc0
            elif nibble0 == 0b0100:
                # tmp <= 0x0f
                mask = 0x80
            else:
                # false
                mask = 0x00

            t0 = (tmp & mask)
            flag = t0 != 0

            print(f"{nibble0:02x}: {continuation1 & 0x3f:02x} & {mask:02x} = {t0:02x} ({flag})")

            return (flag, "len4*")
            

    return (False, "other")


def main():
    failed = 0
    for i, (reason, expected, byte0, byte1) in enumerate(generate_words()):
        result, which = algorithm(byte0, byte1)
        if result != expected:
            #print(reason)
            print(failed, '\t', f'{byte0:02x} {byte1:02x} -> expected={expected}, result={result} {which}')
            print()
            failed += 1


if __name__ == '__main__':
    main()

File: avx512-utf8-to-utf32/validate/avx512-validate-leading-bytes.cpp
// included in avx512-validate-utf8.cpp

/*
This is a direct translation of branchy algorithm from
"Validating UTF-8 In Less Than One Instruction Per Byte".

See the implementation in reference.cpp.

There are following observations:

1. ASCII bytes do not need handling
2. Continuation bytes are rejected instantly
3. For 2-/3-/4-byte chars it's sufficient to check the values
   of leading byte and the first continuation byte. The
   subsequent continuation bytes may have any values

a) For 2-byte chars we have one rule:

- The leading byte must not be 0xc0 nor 0xc1.

b) For 3-byte chars we decide based on the lower nibble of the leading byte:

- 0: valid if (continuation1 & 0x3f) > 0x1f
- 14: valid if (continuation1 & 0x3f) <= 0x1f [negation of the above case]
- 0..13, 14, 15: always valid

We calculate the following expression:

    c = (continuation1 & 0x3f)
    x = (c ^ fixup) > 0x1f

    where `fixup` is a function of nibble value:

        fixup[nibble] = 0  => 0x00
                        14 => 0x40
                        _  => 0xc0

    case 1 fixup = 0x00:
                                                          c ^ fixup > 0x1f
    c <= 0x1f: 0b000x_xxxx ^ 0xb0000_0000 = 0xb000x_xxxx  false
    c >  0x1f: 0b001x_xxxx ^ 0xb0000_0000 = 0xb001x_xxxx  true

    case 2 fixup = 0x40:
                                                          c ^ fixup > 0x1f
    c <= 0x1f: 0b000x_xxxx ^ 0xb0010_0000 = 0xb001x_xxxx  true
    c >  0x1f: 0b001x_xxxx ^ 0xb0010_0000 = 0xb000x_xxxx  false

    case 3 fixup = 0x40:
                                                          c ^ fixup > 0x1f
    c <= 0x1f: 0b000x_xxxx ^ 0xb1100_0000 = 0xb110x_xxxx  true
    c >  0x1f: 0b001x_xxxx ^ 0xb1100_0000 = 0xb111x_xxxx  true


c) For 4-byte chars we decide based on the lower nibble of the leading byte:

- 0: valid if (continuation1 & 0x3f) > 0x0f
- 1, 2, 3: always valid
- 4: valid if (continuation1 & 0x3f) <= 0x0f [negation of case #0]
- 5, 15: always invalid

In this case we build a 2-bit values where one bit stores result of
(continuation1 & 0x3f) > 0x0f and another is negation of that bit.  Then case
#0 is encoded as mask 0b01, case #4 is encoded as 0b10, cases #{1,2,3} as 0b11
and the remaining cases as 0b00. The validity of chars is checked with the
following expression:

    value & mask != 0

To create a 2-bit value we subtract from c = (continuation1 & 0x3f) the number
0x10. If c = 0b0000_xxxx (<= 0x0f) then the result of the subtraction is always
in form 0b1111_xxxx. The two most significant bits encodes condition (c <= 0x0f).
Now we negate 6th bit, and then the these two most significant bits represents
the required combination which is later tested.
*/
__mmask64 avx512_validate_leading_bytes(__m512i leading_bytes, __m512i continuation1, __mmask64 tested_chars) {

    const __m512i nibble0 = _mm512_and_si512(leading_bytes, v_0f);

    // 1. Assume all non-continuation bytes are valid leading bytes
    //    We unset all continuation bytes (0b10xx_xxxx) and ASCII chars
    //    (0b0xxx_xxxx) --- so we looking for 0b11xx_xxxx chars.
    //
    //    (We may mask-out some leading bytes via `tested_chars`).
    __mmask64 valid;
    {
        const __m512i t0 = _mm512_and_si512(leading_bytes, v_c0);
        valid = _mm512_cmpeq_epi8_mask(t0, v_c0);
    }

    __mmask64 _2bytes;
    {
        const __m512i t0 = _mm512_and_si512(leading_bytes, v_e0);
        _2bytes = _mm512_cmpeq_epi8_mask(t0, v_c0);
    }

    __mmask64 _3bytes;
    {
        const __m512i t0 = _mm512_and_si512(leading_bytes, v_f0);
        _3bytes = _mm512_cmpeq_epi8_mask(t0, v_e0);
    }

    __mmask64 _4bytes;
    {
        const __m512i t0 = _mm512_and_si512(leading_bytes, v_f8);
        _4bytes = _mm512_cmpeq_epi8_mask(t0, v_f0);
    }

    // 1. Handle 2-byte chars
    //    Valid if leading byte is greater than 0xc2
    __mmask64 valid_2bytes = _mm512_mask_cmpge_epu8_mask(_2bytes, leading_bytes, v_c2);

    // 4. Handle 3-byte chars
    //    let M = (continuation1 & 03f) > 0x1f
    continuation1 = _mm512_and_si512(continuation1, v_3f);
    __mmask64 valid_3bytes;
    {
        /** pshufb 
            M     = 0b0000_0000 # we test 5th bit
            notM  = 0b0010_0000
            true  = 0b1100_0000

            fixup_lookup = 4 * [
                M,      # 0b0000
                true,   # 0b0001
                true,   # 0b0010
                true,   # 0b0011
                true,   # 0b0100
                true,   # 0b0101
                true,   # 0b0110
                true,   # 0b0111
                true,   # 0b1000
                true,   # 0b1001
                true,   # 0b1010
                true,   # 0b1011
                true,   # 0b1100
                notM,   # 0b1101
                true,   # 0b1110
                true,   # 0b1111
        ] */
        const __m512i fixup_lookup = _mm512_setr_epi64(
            0xc0c0c0c0c0c0c000,
            0xc0c020c0c0c0c0c0,
            0xc0c0c0c0c0c0c000,
            0xc0c020c0c0c0c0c0,
            0xc0c0c0c0c0c0c000,
            0xc0c020c0c0c0c0c0,
            0xc0c0c0c0c0c0c000,
            0xc0c020c0c0c0c0c0
        );

        const __m512i fixup = _mm512_shuffle_epi8(fixup_lookup, nibble0);
        const __m512i t0 = _mm512_xor_si512(fixup, continuation1);
        valid_3bytes = _mm512_mask_cmpgt_epu8_mask(_3bytes, t0, v_1f);
    }

    // 5. Handle 4-byte chars
    __mmask64 valid_4bytes;
    {
        // continuation1 in range [0..63] (0b0000_0000 .. 0b0011_1111)
        //
        // case 1: c1 <= 0xf:  0b0000_xxxx - 0x10 = 0b1111_yyyy
        // case 2: c1  > 0xf:  0b00xx_xxxx - 0x10 = 0b00yy_yyyy
        __m512i t0;
        t0 = _mm512_sub_epi8(continuation1, v_10);
        t0 = _mm512_xor_si512(t0, v_40);
        // bit 7th of c: continuation1 <= 0x0f
        // bit 6th of c: continuation1 >  0x0f

        /** pshufb 
            le_0f   = 0x80  # c1[7] = continuation1 <= 0xf0
            gt_0f   = 0x40  # c1[6] = not c1[7]
            true    = gt_0f | le_0f
            false   = 0x00

            mask_lookup = 4 * [
                gt_0f,  # 0b0000
                true,   # 0b0001
                true,   # 0b0010
                true,   # 0b0011
                le_0f,  # 0b0100
                false,  # 0b0101
                false,  # 0b0110
                false,  # 0b0111
                false,  # 0b1000
                false,  # 0b1001
                false,  # 0b1010
                false,  # 0b1011
                false,  # 0b1100
                false,  # 0b1101
                false,  # 0b1110
                false,  # 0b1111
        ] */
        const __m512i mask_lookup = _mm512_setr_epi64(
            0x00000080c0c0c040,
            0x0000000000000000,
            0x00000080c0c0c040,
            0x0000000000000000,
            0x00000080c0c0c040,
            0x0000000000000000,
            0x00000080c0c0c040,
            0x0000000000000000
        );
        const __m512i mask = _mm512_shuffle_epi8(mask_lookup, nibble0);

        valid_4bytes = _mm512_mask_test_epi8_mask(_4bytes, mask, t0);
    }

    {
        // all: marks all valid leading bytes
        __mmask64 all = _kor_mask64(valid_2bytes, _kor_mask64(valid_3bytes, valid_4bytes));

        // reset leading byte marks from input
        valid = _kxor_mask64(valid, all);
        valid = _kand_mask64(valid, tested_chars);
    }

    return valid == 0;
}
